### DISCUSSION TITLE
Gatsby v4 - search issue - allSitePages data not available for use by plugins/gatsby-config.js

### DISCUSSION THREAD

**Discussion by TristanPeers** [gatsbyjs/gatsby#35142] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

We are trying to upgrade from Gatsby v2 to v4. Our site is a mix of markdown and dynamic data pages which are generated from external data. 
We need to add all the pages to the site search, therefore we need to use `AllSitePages` in our GraphQL queries as opposed to `AllMarkdown/AllMdx`.
We used the Elasticsearch plugin in Gatsby v2, this worked very well. As you can see the page even [lists AllSitePage](https://www.gatsbyjs.com/plugins/@logilab/gatsby-plugin-elasticsearch/) as the query in their example.

In Gatsby v4 however, it seems `AllSitePage` data is no longer available to the plugins in `gatsby-config.js`. We cannot find any mention of this in either the v3 or v4 release notes.

We have tried moving to [Gatsby Plugin Local Search](https://www.gatsbyjs.com/plugins/gatsby-plugin-local-search/) yet this also has the same issue. 

gatsby-config.js:

```module.exports = {
  siteMetadata: {
    title: `new`,
    siteUrl: `https://www.yourdomain.tld`,
  },
  plugins: [
    "gatsby-plugin-mdx",
    {
      resolve: "gatsby-source-filesystem",
      options: {
        name: "pages",
        path: "./src/pages/",
      },
      __key: "pages",
    },
    {
      resolve: "gatsby-plugin-local-search",
      options: {
        name: "pages",
        engine: "flexsearch",
        query: `
          query {
            allSitePage {
              nodes {
                path
              }
            }
          }
        `,
        ref: "path",
        index: ["path"],
        store: ["path"],
        normalizer: ({ data }) =>
          data.allSitePage.nodes.map((node) => ({
            path: node.path,
          })),
      },
    },
  ],
};
```

### Reproduction Link

https://github.com/TristanPeers/gatsby-v4

### Steps to Reproduce

1. Clone repository above
2. Run gatsby develop

### Expected Result

AllSitePage data is parsed and added to search index

### Actual Result

Error shows no data:

```
$ gatsby develop
success compile gatsby files - 0.470s
success load gatsby config - 0.027s
success load plugins - 0.744s
success onPreInit - 0.004s
info One or more of your plugins have changed since the last time you ran
Gatsby. As
a precaution, we're deleting your site's cache to ensure there's no stale data.
success initialize cache - 0.161s
success copy gatsby files - 0.175s
success Compiling Gatsby Functions - 0.276s
success onPreBootstrap - 0.294s
success createSchemaCustomization - 0.010s
success Checking for changed pages - 0.001s
success source and transform nodes - 0.529s
success building schema - 0.312s
warn gatsby-plugin-local-search - The query for index "pages" returned no nodes.
 The index and store will be empty.
success createPages - 0.018s
success createPagesStatefully - 0.112s
info Total nodes: 36, SitePage nodes: 5 (use --verbose for breakdown)
success Checking for changed pages - 0.002s
success write out redirect data - 0.027s
success onPostBootstrap - 0.004s
info bootstrap finished - 6.326s
success onPreExtractQueries - 0.001s
```

### Environment

```shell
System:
    OS: macOS 11.3.1
    CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 16.11.1 - /usr/local/bin/node
    Yarn: 1.22.17 - ~/.homebrew/bin/yarn
    npm: 8.0.0 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Safari: 14.1
  npmPackages:
    gatsby: ^4.9.3 => 4.9.3
    gatsby-plugin-local-search: ^2.0.1 => 2.0.1
    gatsby-plugin-mdx: ^3.9.1 => 3.9.1
    gatsby-source-filesystem: ^4.9.1 => 4.9.1
```

### Config Flags

No

**Comment by LekoArts** [comment]
Hi!

The `SitePage` data is not available in `createPages` lifecycle yet: https://github.com/angeloashmore/gatsby-plugin-local-search/blob/master/src/gatsby-node.ts#L89
Only after `createPages` that information is available.

The https://gitlab.com/logilab/gatsby-plugin-elasticsearch should fundamentally work as it queries that data in `onPostBuild` where it's available. I can't say at the moment why it wouldn't work as I'm not familiar with it. Have you tried debugging it?

**Comment by TristanPeers** [comment]
Hi @LekoArts - many thanks for looking.

I listed that plugin as they show the `AllSitePages` in their documentation pages.

We are actually using [this plugin](https://www.gatsbyjs.com/plugins/@gatsby-contrib/gatsby-plugin-elasticlunr-search/) with V2 and it is working fine reading `SitePage` data.

I have also added this plugin to my [example repo](https://github.com/TristanPeers/gatsby-v4) to demonstrate:

Here is the output from `gatsby develop`:

```
$ gatsby develop
success compile gatsby files - 22.044s
success load gatsby config - 0.036s
success load plugins - 0.905s
success onPreInit - 0.006s
info One or more of your plugins have changed since the last time you ran Gatsby. As
a precaution, we're deleting your site's cache to ensure there's no stale data.
success initialize cache - 0.308s
success copy gatsby files - 0.257s
success Compiling Gatsby Functions - 0.378s
success onPreBootstrap - 0.399s
success createSchemaCustomization - 0.012s
warn The @gatsby-contrib/gatsby-plugin-elasticlunr-search plugin has generated no Gatsby nodes. Do you need it?
success Checking for changed pages - 0.002s
success source and transform nodes - 0.554s
success building schema - 0.466s
warn gatsby-plugin-local-search - The query for index "pages" returned no nodes. The index and store will be empty.
success createPages - 0.031s
success createPagesStatefully - 0.323s
info Total nodes: 39, SitePage nodes: 5 (use --verbose for breakdown)
success Checking for changed pages - 0.001s
success write out redirect data - 0.002s
success onPostBootstrap - 0.006s
info bootstrap finished - 32.360s
success onPreExtractQueries - 0.001s
```

As you can see both plugins don't generate any nodes. In V2 there is no issue and the `SitePage` data is used to generate the search index without a problem. Has there been some lifecycle change from v2 -> v4?

Thanks

**Reply by LekoArts** [reply]
I'm reading the source code of https://github.com/gatsby-contrib/gatsby-plugin-elasticlunr-search/blob/master/src/gatsby-node.js a little bit and it might very be the case that the plugin is not compatible with v4 in its current form. With v4 we changed quite a few things, like introducing Parallelization/Parallel Query Running and using a new datastore (LMDB). With that some things under the hood changed, including when certain data is available and how you can query data.

Sadly I'm not really familiar with what the plugin is doing, this would need more investigation.

**Comment by tony-gutierrez** [comment]
In what case would allSitePages have null for internal->content? I can query fine after page build, but there is no content?
